## **FeedForward Neural Network Propagation for iris dataset**
Feedforward neural networks, or multi-layer perceptrons (MLPs), are what we’ve primarily been focusing on within this article. They are comprised of an input layer, a hidden layer or layers, and an output layer. While these neural networks are also commonly referred to as MLPs, it’s important to note that they are actually comprised of sigmoid neurons, not perceptrons, as most real-world problems are nonlinear. Data usually is fed into these models to train them, and they are the foundation for computer vision, natural language processing, and other neural networks.

### Imporing libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

### Importing dataset

In [2]:
data=pd.read_csv(r'iris_data.csv')
# Viewing the dataset
data.head()

sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

### Pre-Processing dataset

In [3]:
# To check rows and columns
data.shape

(150, 5)

#### Encoding the species

In [4]:
# Converting category value to an integer number using label encoder 
label=LabelEncoder()
label=label.fit_transform(data.iloc[:,-1])     #-1 means the end coulmn ie species column
#label.shape        #it gives no columns and we assume there is 1 column but numpy array finds it difficult so we add a column
label=label.reshape((-1,1))     #-1 says keep original rows comma +1 says make column one
label.shape

(150, 1)

In [5]:
# now we put label values in data
data.iloc[:,-1]=label
data 
# now our species will have values 0,1,2 ie 3 types of species

sepal_length  sepal_width  petal_length  petal_width  species
0             5.1          3.5           1.4          0.2        0
1             4.9          3.0           1.4          0.2        0
2             4.7          3.2           1.3          0.2        0
3             4.6          3.1           1.5          0.2        0
4             5.0          3.6           1.4          0.2        0
..            ...          ...           ...          ...      ...
145           6.7          3.0           5.2          2.3        2
146           6.3          2.5           5.0          1.9        2
147           6.5          3.0           5.2          2.0        2
148           6.2          3.4           5.4          2.3        2
149           5.9          3.0           5.1          1.8        2

[150 rows x 5 columns]

#### Normalizing the dataset

In [6]:
# remember to normalize distribution if required to avoid problems 

In [7]:
#Scaling of attributes (it makes the mean zero)
scaler=StandardScaler()
# scaling because we are preprocessing our dataset ,the first 4  columns has to be scaled
input_float=data.iloc[:,0:4]
# fitting the model
scaler.fit(input_float)
# normalizing the dataset
input_float=scaler.transform(input_float)
# putting back to data
data.iloc[:,0:4]=input_float
# we take the mean value of each column and subtract it with individual vales to make mean zero that is scaling
data.head()
#we make variance 1 by finding standard deviation and divide it by individual value

sepal_length  sepal_width  petal_length  petal_width  species
0     -0.900681     1.032057     -1.341272    -1.312977        0
1     -1.143017    -0.124958     -1.341272    -1.312977        0
2     -1.385353     0.337848     -1.398138    -1.312977        0
3     -1.506521     0.106445     -1.284407    -1.312977        0
4     -1.021849     1.263460     -1.341272    -1.312977        0

#### Creating output columns

In [8]:
# adding three  coloumns so that if species is 0 then flower1 will have value 1 and other two flower columns will have value 0
data.insert(5,"flower1",0)
data.insert(6,"flower2",0)
data.insert(7,"flower3",0)
# visualizing the dataset
data.head()

sepal_length  sepal_width  petal_length  ...  flower1  flower2  flower3
0     -0.900681     1.032057     -1.341272  ...        0        0        0
1     -1.143017    -0.124958     -1.341272  ...        0        0        0
2     -1.385353     0.337848     -1.398138  ...        0        0        0
3     -1.506521     0.106445     -1.284407  ...        0        0        0
4     -1.021849     1.263460     -1.341272  ...        0        0        0

[5 rows x 8 columns]

In [9]:
# if species is 0 make flower 1 as 1 and rest both flowers will be 0
# if species is 1 make flower 2 as 1 and rest both flowers will be 0
# if species is 2 make flower 3 as 1 and rest both flowers will be 0
[rows,columns]=data.shape  
for i in range(rows):      
    if data.iloc[i,4]==0: 
        data.iloc[i,5]=1
    elif data.iloc[i,4]==1:
         data.iloc[i,6]=1
    else:
        data.iloc[i,7]=1
data

sepal_length  sepal_width  petal_length  ...  flower1  flower2  flower3
0       -0.900681     1.032057     -1.341272  ...        1        0        0
1       -1.143017    -0.124958     -1.341272  ...        1        0        0
2       -1.385353     0.337848     -1.398138  ...        1        0        0
3       -1.506521     0.106445     -1.284407  ...        1        0        0
4       -1.021849     1.263460     -1.341272  ...        1        0        0
..            ...          ...           ...  ...      ...      ...      ...
145      1.038005    -0.124958      0.819624  ...        0        0        1
146      0.553333    -1.281972      0.705893  ...        0        0        1
147      0.795669    -0.124958      0.819624  ...        0        0        1
148      0.432165     0.800654      0.933356  ...        0        0        1
149      0.068662    -0.124958      0.762759  ...        0        0        1

[150 rows x 8 columns]

### Creating Neuron Network

<img src='https://i.ibb.co/VHKWhk1/image.png' style='width:30%'>

#### Creating random weights of nurons

In [10]:
np.random.seed(42)
# weights of input - 1st hidden layer nodes
w_i_h1=np.random.rand(4,1) # 4 inputs multiply by weight go to 1st hidden layer neuron
w_i_h2=np.random.rand(4,1) 
# weights of 1st hidden layer to output layer nodes
w_h_o1=np.random.rand(2,1) 
w_h_o2=np.random.rand(2,1)
w_h_o3=np.random.rand(2,1)
# bais for hidden and output layer
bais1=np.random.rand(1)
bais2=np.random.rand(1)

#### Sigmoid function

In [11]:
# creating the sigmoid fucntion
def sigmoid(x):
  return 1/(1 + np.exp(-x))

#### Feeding data to Neural Network

In [12]:
# converting data into np array
inputs = np.array(data)
#inputs.shape
# we need first 4 columns to provide it to neural network so we extract first 4 columns
feature_set = inputs[:,:4]
feature_set.shape

(150, 4)

In [13]:
# input to hidden layer
# feedforward step-1
z2_1 = np.dot(feature_set, w_i_h1) + bais1
z2_2 = np.dot(feature_set, w_i_h2) + bais2
# feedforward step-2
h2_1 = sigmoid(z2_1)
h2_2 = sigmoid(z2_2)

In [14]:
# hidden to output layer
h2 = np.append(h2_1, h2_2, axis = 1)
z3_1 = np.dot(h2, w_h_o1)
z3_2 = np.dot(h2, w_h_o2)
z3_3 = np.dot(h2, w_h_o3)
o1 = sigmoid(z3_1)
o2 = sigmoid(z3_2)
o3 = sigmoid(z3_3)

#### Calculating Error

In [15]:
# error calculation 
fl1 = np.array(data.iloc[:,5])
fl1 = fl1.reshape((-1,1))
fl2 = np.array(data.iloc[:,6])
fl2 = fl2.reshape((-1,1))
fl3 = np.array(data.iloc[:,7])
fl3 = fl3.reshape((-1,1))

error1 = (o1 - fl1)
error2 = (o2 - fl2)
error3 = (o3 - fl3)

error1_square = np.multiply(error1, error1)
error2_square = np.multiply(error2, error2)
error3_square = np.multiply(error3, error3)

error = error1_square + error2_square + error3_square
print(f'Total Error : {np.sum(error)}')

Total Error : 144.83471642466847


### ***Conclusion*** :
1. Iris dataset of 150 samples is used for feedforward network. Initial weights are generated using randomizer, since back propogation is not used the error is calculated to be 144.83
2. If we change the weight using randomizer the error will change